### PR TITLE
SRVKP-4523: add alert,panels for tekton-pipeline-controllers deadlocks and rapid restarts

### PR DESCRIPTION
**Pull request by gabemontero**
I'll add the panels once https://github.com/redhat-appstudio/infra-deployments/pull/4105 and I can test the changes against app-sre staging

this also replaces https://github.com/redhat-appstudio/o11y/pull/297 as I am pulling in the rest of the alerts I have planned for SRVKP-4523 @zanssa 

Signed-off-by: Gabe Montero <gmontero@redhat.com>

rh-pre-commit.version: 2.3.0
rh-pre-commit.check-secrets: ENABLED

### CODE REVIEW

**Comment by gabemontero** [comment]
@amisstea @yftacherzog PTAL but FYI I will be adding panels after https://github.com/redhat-appstudio/infra-deployments/pull/4105 merges

I don't have permission to request reviewers via github

thanks

**Comment by gabemontero** [comment]
> @gabemontero please add some unit tests for these new alerting rules.

shoot have those but forgot to add the new file to git ... just a sec @amisstea

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (gabemontero) [inline test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:7]: still need to create the deadlock tests @amisstea but I was going to pattern those off the restart one, so let's see if you agree with what I did here and then I'll replicate

thanks

**Comment by gabemontero** [comment]
I'll then also include the panels now that you've merged https://github.com/redhat-appstudio/infra-deployments/pull/4105 @amisstea thanks

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (amisstea) [inline rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:91]:
```suggestion
              Tekton pipeline controller on cluster {{ $labels.source_cluster }} has appeared deadlocked on {{ $value }} taskruns.
```

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (amisstea) [inline test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:17]: This input series is identical to the one above. Time series should be unique when comparing all labels so I think this should be removed.

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (amisstea) [inline test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:21]: Another duplicate time series that should probably be removed.

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (amisstea) [inline test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:25]: This time series might replace the one before it. Which one should we keep?

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (gabemontero) [reply test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:17]: yep forgot how the `+` operator worked for a sec there   .... updating

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (gabemontero) [reply test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:25]: think it makes sense to remove both 300 and 1000 actually

**Comment by gabemontero** [comment]
updates pushed @amisstea thanks

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (gabemontero) [reply test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:17]: left the time series with the pod label diff and different incrementation to validate this one is not counted  ... but let me know @amisstea if I'm still not entirely in sync with how the unit test time series work

Comment (amisstea) [reply test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:17]: This simpler set of time series looks great!

File: test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml
Comment (yftacherzog) [inline test/promql/tests/data_plane/pipeline_deadlock_crashloop_test.yaml:58]: add newline

**Review by yftacherzog — APPROVED** [review]
lgtm

**Comment by gabemontero** [comment]
@amisstea @yftacherzog best as I can tell the EC CI failure is unrelated to this change.  I'll attach the log.  After getting a chance to look at the log, let me know if you all agree.

Thanks
[verify.log](https://github.com/user-attachments/files/16285622/verify.log)

**Comment by amisstea** [comment]
@gabemontero please rebase your branch and the EC errors should be resolved. There were recent changes to the `EnterpriseContractPolicy` which subsequently required updates to our pipeline definitions (to adopt OCI trusted artifacts).

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (zanssa) [inline rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:52]: I have a concern here about the time window. I am afraid that there would be false positive alerts. IIRC, there are 2 controller pods in a cluster. If these both are restarted while doing upgrade/deployment, I shouldn't get PD alert in that case. 

Can I ask if it would be better to adjust the time window & thresholds? 

Also, does the severity level correct for all envs? I am just wondering if this is would be applied for staging clusters and both envs (prod & stage) have same severity levels.

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (zanssa) [inline rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:64]: I should also point out the importance of having a SOP for each SLO alert. It would be great to work on that as soon as possible :)

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (gabemontero) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:52]: Can you give me suggestions for the time windows and thresholds @zanssa ?
You are correct about 2 controller pods existing in the konflux clusters.

WRT varying severity AFAIK I cannot distinguish between stage and prod with these alertrs and their severity.  Do I have that correct @amisstea ?

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (gabemontero) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:64]: Yes we'll be updating the SOPs i.e. https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/data/services/stonesoup/pipeline-service/slo-sli-document.yml?ref_type=heads as part of SRVKP-4523 and SRVKP-4524 as well as planning to add the app-sre panels in this PR per my earlier comments.

We'll also be updating https://gitlab.cee.redhat.com/konflux/docs/sop/-/tree/main/pipeline-service/slos?ref_type=heads adding new markdowns for the new alerts

**Comment by gabemontero** [comment]
/hold

rebased @amisstea thanks

@zanssa hopefully either the prow hold command above or adding the WIP prefix to the PR should prevent any merging before the panel are updated and the SOP etc. updates have started

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (amisstea) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:52]: The severity levels are unchanged across envs for all our alerts right now. We haven't used them for anything other than display purposes so far.

Comment (zanssa) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:52]: what is your thoughts if we modify the time window for `increase` to be 5m instead of 2m ? 

Before increasing the threshold, I would like to test this and for now I am not aware of an env where I can play with the query. It would be easy to modify that later and I don't wanna block merging this. 

Thanks a lot for working and coordinating on this!

Comment (gabemontero) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:52]: sure @zanssa I'll tweak the timewindow to `5m` and see if I have to adjust the test cases any

The query is now possible from https://grafana.stage.devshift.net/d/rhtap-slos/rhtap-slos?orgId=1&refresh=1h if that is place where you are comfortable performing test queries

Also, if you have login credentials to the konflux clusters themselves , you can go to the OCP console's Observe tab and try there.

Comment (gabemontero) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml]: updated timewindow in a separate commit for now @zanssa  ... ptal when you get the chance

Comment (amisstea) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml]: You can also use the explore view from AppSRE's grafana. Here's the prod instance (more stable): https://grafana.app-sre.devshift.net/explore

Select either of the 'rhtap-observatorium-production` or `rhtap-observatorium-stage` datasources depending on the env you want to query.

Comment (zanssa) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml]: Cool. Thank you @gabemontero & @amisstea. This is helpful. I'll definitely explore these options later. 

While exploring, I don't really wanna keep this on hold, and since we have increased the time window for `increase`, we can merge this and modify later if needed.

Comment (gabemontero) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml]: OK @zanssa I have removed the WIP prefix, though it does not appear to be enabled for this repo (and I can't add labels).

Comment (zanssa) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml]: Feel free to resolve the conversations as I can't do that to unblock the merge.

File: dashboards/grafana-dashboard-rhtap-slos.configmap.yaml
Comment (amisstea) [inline dashboards/grafana-dashboard-rhtap-slos.configmap.yaml:1]: @gabemontero can you please attach some screenshots to show the end result? It's a bit tricky to review this chunk of json :sweat_smile:.

**Comment by gabemontero** [comment]
container-restart-panel

screen shot @amisstea ^^

File: dashboards/grafana-dashboard-rhtap-slos.configmap.yaml
Comment (gabemontero) [reply dashboards/grafana-dashboard-rhtap-slos.configmap.yaml:1430]: ack .... also trying to clean up all the diffs for sure before the new new panels I added and hopefully only updating graph positioning for any panels after

Comment (gabemontero) [reply dashboards/grafana-dashboard-rhtap-slos.configmap.yaml:1430]: the yaml lint output is not too helpful, and I get different results locally .... ah, python dependency version skew :-)

**Comment by gabemontero** [comment]
ok the `?w=` file diff is much cleaner, but it is still failing the yaml lint @amisstea 

I've copied lines 725 to 841 (the contiguous chunk for the new panels) to a separate yaml file and have made note of all the gridPos y settings in the lines after the new panels (i.e. panels have been "bumped down" the browser window)

gridPos y settings

848 - 26
865 - 27
934 - 27
977 - 34
1158 - 41
1197 - 27
1214 - 28
1281 - 28
1354 - 28

I'll save my current file, then revert the change in the branch and try simply adding the subset yaml in the right spot and then update the y positions at lines 848, 865, etc. tomorrow @amisstea 

I suspect the lint is complaining about spaces vs. tabs but it is unclear from the messages ... if you happen to have any experience around this of course let me know.

thanks

File: dashboards/grafana-dashboard-rhtap-slos.configmap.yaml
Comment (amisstea) [inline dashboards/grafana-dashboard-rhtap-slos.configmap.yaml:1445]:
```suggestion
      "timezone": "",
      "title": "RHTAP SLOs",
      "uid": "rhtap-slos",
      "version": 4,
      "weekStart": ""
    }
```

This should fix the yamllint errors.

Comment (gabemontero) [reply dashboards/grafana-dashboard-rhtap-slos.configmap.yaml]: yep yaml lint passes now @amisstea thanks so much  

I've squashed / signed / repushed the commits so the gitlint passes

**Review by amisstea — APPROVED** [review]
LGTM

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (gabemontero) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:64]: @zanssa - step one in the SOP / SLO updates - https://gitlab.cee.redhat.com/konflux/docs/sop/-/merge_requests/163

I'll start on the associated update to https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/data/services/stonesoup/pipeline-service/slo-sli-document.yml?ref_type=heads shortly

Comment (gabemontero) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml:64]: step two is https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/112711 @amisstea @zanssa 

then I will update https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/data/services/stonesoup/pipeline-service/slo-sli-document.yml?ref_type=heads

thanks

File: rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml
Comment (zanssa) [reply rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml]: @amisstea do I need to have a specific permission to be able to use the explore view from AppSRE's grafana? If I tried to open the link, I would get `Login failed. Failed to get token from provider` - I clicked on sign in with SSO, and I got the same error msg.

**Pull request merged**
